File: listas/05.programacao-funcional/snake-clj/snake-game.clj
;; Snake Game Functional in Clojure - Caio Cesar Vieira Cavalcanti - 123110825 - 01/05/2025
;; Problema - A cobra não se movimenta a cada tick automaticamente


(ns snake-game
  (:require [clojure.string :as str])
  (:import [org.jline.terminal TerminalBuilder Terminal]
           [org.jline.reader.impl LineReaderImpl]))

;; ---------------------------------------------------------------------
;; Constants and utility functions
;; ---------------------------------------------------------------------

(def WIDTH 20)
(def HEIGHT 10)
(def TICKRATE 300)

(def directions {:UP [0 -1]
                 :DOWN [0 1]
                 :LEFT [-1 0]
                 :RIGHT [1 0]})

(def opposites {:UP :DOWN, :DOWN :UP, :LEFT :RIGHT, :RIGHT :LEFT})

(def terminal (-> (TerminalBuilder/builder)
                  (.jna true)
                  (.system true)
                  (.build)))

(def reader (.reader terminal))

(defn set-raw-mode []
  (.enterRawMode terminal))

(defn restore-mode []
  (.close terminal))

(defn wait [ms]
  (Thread/sleep ms))

(defn clear-screen []
  (print "\033[H\033[2J")
  (flush))

;; Random function
(defn rand-pos [snake]
  (let [pos {:x (rand-int WIDTH)
             :y (rand-int HEIGHT)}]
    (if (some #(= pos %) snake)
      (recur snake)
      pos)))

;; ---------------------------------------------------------------------
;; Snake functions
;; ---------------------------------------------------------------------


(defn move-head [head dir]
  (let [[dx dy] (get directions dir)]
    {:x (+ (:x head) dx)
     :y (+ (:y head) dy)}))

(defn collision? [pos snake]
  (or (< (:x pos) 0) (>= (:x pos) WIDTH)
      (< (:y pos) 0) (>= (:y pos) HEIGHT)
      (some #(= pos %) snake)))

(defn update-snake [snake head ate?]
  (if ate?
    (cons head snake)
    (cons head (butlast snake))))

(defn next-state [state]
  (let [head (move-head (first (:snake state)) (:direction state))
        ate? (= head (:food state))
        new-snake (update-snake (:snake state) head ate?)
        dead? (collision? head (:snake state))]
    {:snake new-snake
     :direction (:direction state)
     :food (if ate? (rand-pos new-snake) (:food state))
     :game-over? dead?
     :history (conj (:history state) state)}))

;; ---------------------------------------------------------------------
;; Render terminal
;; ---------------------------------------------------------------------

(defn render-cell [snake food x y]
  (cond
    (= {:x x :y y} (first snake)) "S"
    (some #(= {:x x :y y} %) (rest snake)) "s"
    (= {:x x :y y} food) "*"
    :else "-"))

(defn render [state]
  (clear-screen)
  (doseq [y (range HEIGHT)]
    (println (apply str (for [x (range WIDTH)] (render-cell (:snake state) (:food state) x y))))))

;; ---------------------------------------------------------------------
;; Input - Keyboard WASD directions
;; ---------------------------------------------------------------------

(defn read-key []
  (let [key (.read reader)]
    (when (not= key -1) ;; -1 means no key pressed
      key)))

(defn key->dir [k]
  (case k
    119 :UP
    115 :DOWN
    97 :LEFT
    100 :RIGHT
    nil))

(defn update-dir [direction input-dir]
  (if (and input-dir (not= (opposites input-dir) direction))
    input-dir
    direction))

;; ---------------------------------------------------------------------
;; Replay
;; ---------------------------------------------------------------------

(defn replay [history]
  (doseq [s history]
    (render s)
    (wait 200))
  (println "Replay finished")
  (println "Thanks for playing the Snake Game by Caio Cesar s2"))

;; ---------------------------------------------------------------------
;; Game loop
;; ---------------------------------------------------------------------

(defn loop-game [state]
  (render state)
  (if (:game-over? state)
    (do
      (println "\nGame Over! Replay in 3s")
      (wait 3000)
      (replay (:history state)))
    (let [input (read-key)
          new-dir (if input
                    (update-dir (:direction state) (key->dir input))
                    (:direction state))
          nextS (next-state (assoc state :direction new-dir))]
      (wait TICKRATE)
      (recur nextS))))


;; ---------------------------------------------------------------------
;; Main
;; ---------------------------------------------------------------------

(defn -main []
  (set-raw-mode)
  (try
    (let [initial {:snake [{:x 5 :y 5}]
                   :direction :RIGHT
                   :food {:x 10 :y 5}
                   :game-over? false
                   :history []}]
      (loop-game initial))
    (finally
      (restore-mode))))

(-main)

File: listas/05.programacao-funcional/snake-js/snake-game.js
/* Snake Game Functional in JavaScript - Caio Cesar Vieira Cavalcanti - 123110825 - 30/4/2025 */

// ---------------------------------------------------------------------
/* States and constants initialization */
// ---------------------------------------------------------------------

const readline = require('readline');
const { stdin, stdout } = process;

const WIDTH = 20;
const HEIGHT = 10;
const TICKRATE = 300;

const directions = {
	UP: [0, -1],
	DOWN: [0, 1],
	LEFT: [-1, 0],
	RIGHT: [1, 0]
};

const opposites = {
	UP: 'DOWN',
	DOWN: 'UP',
	LEFT: 'RIGHT',
	RIGHT: 'LEFT'
};

const initialState = {
	snake: [ { x: 5, y: 5 }],
	direction: 'RIGHT',
	food: { x: 10, y: 5 },
	gameOver: false,
	history: []
};


// ---------------------------------------------------------------------
/* Functions and Game Initialization */
// ---------------------------------------------------------------------

function moveHead(head, dir) {
	const [dx, dy] = directions[dir];
	return { x: head.x + dx, y: head.y + dy };
}

function isCollision(pos, snake) {
	return (
		pos.x < 0 || pos.x >= WIDTH ||
		pos.y < 0 || pos.y >= HEIGHT ||
		snake.some(seg => seg.x === pos.x && seg.y === pos.y)
	);
}

function generateFood(snake) {
	let pos;
	do {
		pos = { x: Math.floor(Math.random() * WIDTH),
			y: Math.floor(Math.random() * HEIGHT) };
	} while (snake.some(seg => seg.x === pos.x && seg.y === pos.y));
	return pos;
}

function updateState(prev) {
	const head = moveHead(prev.snake[0], prev.direction);
	const ate = head.x === prev.food.x && head.y === prev.food.y;
	const newSnake = [head, ...prev.snake.slice(0, ate ? undefined : -1)];
	const dead = isCollision(head, prev.snake);

	return {
		...prev,
		snake: newSnake,
		food: ate ? generateFood(newSnake) : prev.food,
		gameOver: dead,
		history: [...prev.history, prev]
	};
}

/* Terminal */
function render(state) {
	stdout.write('\x1B[H');
	for (let y = 0; y < HEIGHT; y++) {
		let line = '';
		for(let x = 0; x < WIDTH; x++) {
			const isHead = x === state.snake[0].x && y === state.snake[0].y;
			const isBody = state.snake.slice(1).some(s => s.x === x && s.y === y);
			const isFood = state.food.x === x && state.food.y === y;
			line += isHead ? 'S' : isBody ? 's' : isFood ? '*' : '-';
		}
		console.log(line);
	}
}

/* Input - Keyboard WASD */
function startInput(directionChange) {
	stdin.setRawMode(true);
	stdin.resume();
	stdin.setEncoding('utf8');
	stdin.on('data', key => {
		if (key === '\u0003') process.exit(); // Ctrl + C - End game
		const dirMap = { w: 'UP', a: 'LEFT', s: 'DOWN', d: 'RIGHT' };
		const newDir = dirMap[key];
		if (newDir) directionChange(newDir);
	});
}

/* Game Loop */
function gameLoop(initialState) {
	console.clear();

	let currentState = initialState;
	let nextDirection = currentState.direction;

	startInput(newDir => {
		if (opposites[newDir] !== currentState.direction) {
			nextDirection = newDir;
		}
	});

	const interval = setInterval( () => {
		currentState = {
			...currentState,
			direction: nextDirection
		};

		currentState = updateState(currentState);
		render(currentState);
		
		if(currentState.gameOver) {
			clearInterval(interval);
			console.log('\nGame Over! Well played');
			console.log('Replay in 3 seconds...');
			setTimeout( () => {
				console.clear();
				replay(currentState.history);
			}, 3000);
		}
	}, TICKRATE);
}

/* Replay */
function replay(history) {
	let i = 0;
	const replayInterval = setInterval( () => {
		if(i >= history.length) {
			clearInterval(replayInterval);
			console.log('Replay finished');
			console.log('Thanks for playing the Snake Game by Caio Cesar s2');
			process.exit();
		} else {
			render(history[i]);
			i++;
		}
	}, 300);
}


/* Start the game */
gameLoop(initialState);
